test/common/upstream/maglev_lb_test.cc

```diff
@@ -224,24 +224,60 @@ TEST_F(MaglevLoadBalancerTest, LocalityWeightedGlobalPanic) {
   // maglev: i=5 host=127.0.0.1:91
   // maglev: i=6 host=127.0.0.1:91
   // maglev: i=7 host=127.0.0.1:90
-  // maglev: i=8 host=127.0.0.1:91
+  // maglev: i=8 host=127.0.0.1:90
   // maglev: i=9 host=127.0.0.1:91
-  // maglev: i=10 host=127.0.0.1:91
+  // maglev: i=10 host=127.0.0.1:90
   // maglev: i=11 host=127.0.0.1:91
-  // maglev: i=12 host=127.0.0.1:91
+  // maglev: i=12 host=127.0.0.1:90
   // maglev: i=13 host=127.0.0.1:90
   // maglev: i=14 host=127.0.0.1:91
   // maglev: i=15 host=127.0.0.1:90
   // maglev: i=16 host=127.0.0.1:91
   LoadBalancerPtr lb = lb_->factory()->create();
-  const std::vector<uint32_t> expected_assignments{1, 0, 0, 1, 0, 1, 1, 0, 1,
-                                                   1, 1, 1, 1, 0, 1, 0, 1};
+  const std::vector<uint32_t> expected_assignments{1, 0, 0, 1, 0, 1, 1, 0, 0,
+                                                   1, 0, 1, 0, 0, 1, 0, 1};
   for (uint32_t i = 0; i < 3 * expected_assignments.size(); ++i) {
     TestLoadBalancerContext context(i);
     EXPECT_EQ(host_set_.hosts_[expected_assignments[i % expected_assignments.size()]],
               lb->chooseHost(&context));
   }
 }
 
+// Given extremely lopsided locality weights, and a table that isn't large enough to fit all hosts,
+// expect that the least-weighted hosts appear once, and the most-weighted host fills the remainder.
+TEST_F(MaglevLoadBalancerTest, LocalityWeightedLopsided) {
+  host_set_.hosts_.clear();
+  HostVector heavy_but_sparse, light_but_dense;
+  for (uint32_t i = 0; i < 1024; ++i) {
+    auto host(makeTestHost(info_, fmt::format("tcp://127.0.0.1:{}", i)));
+    host_set_.hosts_.push_back(host);
+    (i == 0 ? heavy_but_sparse : light_but_dense).push_back(host);
+  }
+  host_set_.healthy_hosts_ = {};
+  host_set_.hosts_per_locality_ = makeHostsPerLocality({heavy_but_sparse, light_but_dense});
+  host_set_.healthy_hosts_per_locality_ = host_set_.hosts_per_locality_;
+  host_set_.locality_weights_ = makeLocalityWeights({127, 1});
+  host_set_.runCallbacks({}, {});
+  init(MaglevTable::DefaultTableSize);
+
+  LoadBalancerPtr lb = lb_->factory()->create();
+
+  // Populate a histogram of the number of table entries for each host...
+  uint32_t counts[1024] = {0};
+  for (uint32_t i = 0; i < MaglevTable::DefaultTableSize; ++i) {
+    TestLoadBalancerContext context(i);
+    uint32_t port = lb->chooseHost(&context)->address()->ip()->port();
+    ++counts[port];
+  }
+
+  // Each of the light_but_dense hosts should appear in the table once.
+  for (uint32_t i = 1; i < 1024; ++i) {
+    EXPECT_EQ(1, counts[i]);
+  }
+
+  // The heavy_but_sparse host should occupy the remainder of the table.
```

Comment:
Is it worth adding to the documentation some discussion on the effects of granularity as the number of light weighted hosts approaches the size of the Maglev table? Basically, you get some kind of dilation effect which is a bit unintuitive.

Reply:
Yeah, probably worth documenting. But it makes me wonder about two things:

* How will users know if they're experiencing this behavior?
* What can they do about it?

It seems like, if we really want to pursue that, I should probably add a couple gauges here, like I did in ring hash: `min_hashes_per_host` and `max_hashes_per_host`. And table size should be configurable (we'll pick the next prime number >= the parameter's value). Do you think this is worth doing?

Reply:
Sure, these are good metrics!

Reply:
I don't think it's worth making table size configurable in the context of this PR though, maybe just open an issue; when someone wants it, we can add :)

Reply:
Done!